accounting/invoice/tests/test_choices.py

```diff
@@ -35,3 +43,97 @@ def test_approved_choices(self):
     def test_not_approved_choices(self):
         """ The not-approved choices list matches those choices that actually aren't counted as being approved. """
         self.assertEqual(choices.InvoiceApproval.not_approved_choices(), ['not_approved'])
+
+
+@ddt.ddt
+class InvoiceNumberingSchemeTestCase(TestCase):
+    """ Test cases for `choices.InvoiceNumberingScheme`. """
+
+    @freezegun.freeze_time(NOW)
+    def setUp(self):
+        """ Setup test objects. """
+        super().setUp()
+
+    def test_default_default(self):
+        """ The default value of the `default` choice corresponds to the correct year and month. """
+        self.assertEqual(choices.InvoiceNumberingScheme.default_default(), '2018-01')
+
+    def test_default_number(self):
+        """ The default value of the `number` choice corresponds to the correct integer. """
+        self.assertEqual(choices.InvoiceNumberingScheme.default_number(), '1')
+
+    def test_default_year_month_number(self):
+        """ The default value of the `year_month_number` choice corresponds to the correct year, month,
+        and number combo. """
+        self.assertEqual(choices.InvoiceNumberingScheme.default_year_month_number(), '2018-01-1')
+
+    def test_default_opencraft_year_month(self):
+        """ The default value of the `opencraft_year_month` choice corresponds to the correct prefix and
+        year, month combo. """
+        self.assertEqual(choices.InvoiceNumberingScheme.default_opencraft_year_month(), 'OC-2018-01')
+
+    def test_default_opencraft_number(self):
+        """ The default value of the `opencraft_number` choice corresponds to the correct prefix and number combo. """
+        self.assertEqual(choices.InvoiceNumberingScheme.default_opencraft_number(), 'OC-1')
+
+    @ddt.data(
+        ('2018-01', '2018-02'),
+        ('2018-12', '2019-01'),
+    )
+    @ddt.unpack
+    def test_increment_default(self, value, incremented_value):
+        """ Incrementing the `default` choice gives the correct year and month, including for wrap-arounds. """
+        self.assertEqual(choices.InvoiceNumberingScheme.increment_default(value), incremented_value)
+
+    @ddt.data(
+        ('3', '4'),
+        ('000003', '4'),
+    )
+    @ddt.unpack
+    def test_increment_number(self, value, incremented_value):
+        """ Incrementing the `number` choice gives the correct number, including with leading 0s. """
+        self.assertEqual(choices.InvoiceNumberingScheme.increment_number(value), incremented_value)
+
+    @ddt.data(
+        ('2018-01-4', '2018-02-5'),
+        ('2018-12-10', '2019-01-11'),
```

Comment:
@UmanShahzad I'm assuming that for the year-month-number scheme, we might get values like `2018-05-37` or even `2018-05-257`, right? Maybe we can make that more visible by including such a value in the data defined here.

Reply:
:+1: